AIPCC-632: Add ModelCar OCI TA task

Victoremepunto opened the pull request:
Add a ModelCar task to build images in Konflux

This depends on https://github.com/konflux-ci/build-tasks-dockerfiles/pull/252 for the SBOM script

- closes [AIPCC-632](https://issues.redhat.com//browse/AIPCC-632)

ralphbean commented on task/modelcar-oci-ta/0.1/modelcar-oci-ta.yaml:
```diff
@@ -0,0 +1,178 @@
+
+      - name: sbom-generate
+#              image: quay.io/konflux-ci/sbom-utility-scripts@sha256:1939901046f2ec0afda6d48f32dc82f991d9a4e2b4b4513635b9c79e3d4c2872
+        image: quay.io/vmugicag/build-tasks-dockerfiles@sha256:4d0d0095a04e95c5f9451d1d1ddc7908b9174ac53f9925484fb09f17e9299472
```
This line, obviously, should not be merged. It depends on https://github.com/konflux-ci/build-tasks-dockerfiles/pull/252 being merged first before Victor can replace the image ref here.

ralphbean commented on task/modelcar-oci-ta/0.1/modelcar-oci-ta.yaml:
```diff
@@ -0,0 +1,178 @@
+          --output-file sbom.json \
+
+          #TODO: remove debug:
+          cat sbom.json
```
Remove debug?

Victoremepunto replied:
yup! done

ralphbean commented:
@konflux-ci/build-maintainers, this is the task that works around the 200gb limit we ran into with buildah, where the content needed to be copied 3 times? This `olot` tool enables efficiently layering that model data on top of an existing base image.

mmorhun commented on task/olot-oci-ta/0.1/olot-oci-ta.yaml:
```diff
@@ -0,0 +1,339 @@
+    app.kubernetes.io/version: "0.1"
+    build.appstudio.redhat.com/build_type: oci-artifact
+spec:
+  description: Given a file in the user's source directory, copy content from
```
Would be nice to have more detailed description.

Victoremepunto replied:
I expanded the task's `description`

tarilabs commented on task/modelcar-oci-ta/0.1/README.md:
```diff
@@ -0,0 +1,40 @@
+(https://github.com/containers/olot). The process consists on extracting the files from the provided
+Model OCI artifact reference, and putting the layers on top of the provided Base Image.
+
+This operation requires about 2 times the size of the model files' size (as opposed to using a tool
+like podman or buildah where the requirement is of about 2-3 times the model files size), allowing
+to build ModelCar images with lower disk footprint during the build.
```
```suggestion
This operation requires about 2 times the size of the model files' size (as opposed to using a tool
like podman or buildah where the requirement is of about 2-3 times the model files size), allowing
to build ModelCar images with lower disk footprint during the build.

Future optimisations might eventually leverage the capability of OLOT to remove each file once
layered on top of the base image, using the [`--remove-originals` feature flag](https://github.com/containers/olot/pull/17). 
```

tarilabs reviewed (commented):
hi, optional comments for future iterations, hth!

ralphbean commented on task/modelcar-oci-ta/0.1/modelcar-oci-ta.yaml:
```diff
@@ -0,0 +1,178 @@
+          set -eu
+          set -o pipefail
+
+          pip install olot==${OLOT_VERSION}
```
This is a good catch.

Ignoring this `pip install ...` for a sec, I think the approach to hermeticity here is similar to the `oci-copy` task: the user cannot submit open-ended build scripts here. They submit 1) a base image pullspec and 2) a model pullspec and the modelcar task uses only those ingredients to produce the modelcar image. Network-restricted hermeticity is not required, because the nature of the task doesn't permit "random" network requests that aren't tracked in the SBOM.

Back to this `pip install ...`, we should get this _into_ the task image and build that task image hermetically so that this task doesn't have to install it at task run time.

Victoremepunto replied:
first iteration https://github.com/Victoremepunto/olot-container / image at quay.io/vmugicag/olot:latest

Happy to transfer this to Konflux org when ready

ralphbean replied:
> I'll follow up that work - is that okay to leave it outside of this one or would we want to gate this work on it instead ?

In my opinion, it is okay to merge it with the existing `pip install ..` present, as long as they have our commitment to follow up and replace it with a hermetically built `olot` image. I will defer to @mmorhun and other build-definitions maintainers for a decision on this.

ralphbean commented:
@Victoremepunto https://github.com/konflux-ci/build-tasks-dockerfiles/pull/252 is merged. You should be able to update the `image` ref here for that shortly.

Victoremepunto commented on task/modelcar-oci-ta/0.1/modelcar-oci-ta.yaml:
```diff
@@ -0,0 +1,178 @@
+        script: |
+          #!/bin/bash
+          set -eu
+          set -o pipefail              
```
done, thanks !

Victoremepunto commented:
> @Victoremepunto [konflux-ci/build-tasks-dockerfiles#252](https://github.com/konflux-ci/build-tasks-dockerfiles/pull/252) is merged. You should be able to update the `image` ref here for that shortly.

The PR should be ready for review, I've updated it to use the upstream image with the SBOM script, and fixed all other linting issues.

ralphbean commented on CODEOWNERS:
```diff
@@ -121,6 +121,9 @@
 /task/oci-copy-oci-ta   @ralphbean
 
+# renovate groupName=modelcar-oci
+/task/modelcar-oci-ta   @Victoremepunto
```
I can help co-maintain this.

```suggestion
/task/modelcar-oci-ta   @Victoremepunto @ralphbean 
```

ralphbean commented:
/ok-to-test

ralphbean commented:
@konflux-ci/build-maintainers any further review needed here?

tarilabs reviewed (commented):
for your consideration, please; thank you

tarilabs commented:
thank you @Victoremepunto

ralphbean commented:
/ok-to-test

chmeliik reviewed (dismissed):
LGTM, optional nitpicks

ralphbean commented on CODEOWNERS:
```diff
@@ -121,6 +121,9 @@
 /task/oci-copy-oci-ta   @ralphbean
 
+# renovate groupName=modelcar-oci
+/task/modelcar-oci-ta   @Victoremepunto @ralphbean 
```
I just sent him an invite now -> https://github.com/orgs/konflux-ci/people/pending_invitations

Victoremepunto replied:
Accepted, thanks !

chmeliik replied:
Added to `build-definition-contributors`, the CODEOWNERS file is now valid :+1:

chmeliik commented:
@Victoremepunto if you're going to be making changes here, please consider squashing the commits. We prefer having a meaningful commit history

Victoremepunto commented:
> @Victoremepunto if you're going to be making changes here, please consider squashing the commits. We prefer having a meaningful commit history

I've added changes to address the latest comments about the model secret in f1ee686

About the commit history: I was expecting a squash at merge time, leaving this not so meaningful commits for making the reviewer's job easier. 

I can squash them before merging, I presume at least checking this last change is useful in some way.

@chmeliik please, could you review before doing the final push with the squashed history ?
